### PR TITLE
Allow force run app on cloud if loading locally errors

### PR DESCRIPTION
**Pull request by manskx**
## What does this PR do?

User story:
As a user, I want to be able to run lightning apps on the cloud and I don't need to install the dependencies locally. 

If the app() cannot be loaded locally, it shows the following (the message needs to be revisited to tell the user it's better to save debugging time torun the app locally before running on cloud): 

``` 
"Could not load the app() locally. Starting the app directly on the cloud."
```

Limitations:
- features like defining the flow cloudconfig might not be supported/
- DO NOT MERGE until the lightning backend supports dynamic flow server loading

TODO:

Add e2e test for this feature or add it to one of the apps.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

**Comment by nicolai86** [comment]
isn't 
>features like defining the flow cloudconfig might not be supported/

a blocker for co-locating work and flow on the same nodes?

**Comment by manskx** [comment]
> isn't
> 
> > features like defining the flow cloudconfig might not be supported/
> 
> a blocker for co-locating work and flow on the same nodes?

It depends on how it is implemented, if it requires the load of the app() before running on the cloud then yes it won't be compatible with this PR. I believe that ideally we should try to avoid loading app() as much as we can.

**Comment by manskx** [comment]
This message could be improved. 
`Could not load the app() locally. Starting the app directly on the cloud.`

we want to inform the user that it's better to make sure the app runs locally first before running on cloud because it might contain errors. 

@edenafek what do you think ?

File: src/lightning_app/runners/cloud.py [hunk 1]
```diff
@@ -463,6 +465,30 @@ def _project_has_sufficient_credits(self, project: V1Membership, app: Optional[L
...
+            logger.info("Could not load the app() locally. Starting the app directly on the cloud.")
+            # we want to format the exception as if no frame was on top.
+            exp, val, tb = sys.exc_info()
+            listing = traceback.format_exception(exp, val, tb)
+            # remove the entry for the first frame
+            del listing[1]
+            from lightning_app.testing.helpers import EmptyFlow
+
+            # Create a mocking app.
+            app = LightningApp(EmptyFlow())
```
Comment (tchaton) [inline src/lightning_app/runners/cloud.py:484]: This is too much of a hack and will cause `flow_cloud_compute` not to be detected correctly.

Let's investigate dependencies patching as discussed earlier @manskx, so no `ModuleNotFoundError` is triggered.

**Review by tchaton — CHANGES_REQUESTED** [review]

**Review by tchaton — APPROVED** [review]
Approving for now, but we would need a smarter mechanism to better collect the specs.

File: src/lightning_app/runners/cloud.py (hunk 1 above)
Comment (manskx) [reply src/lightning_app/runners/cloud.py:484]: Created an issue to do it in a following PR: https://github.com/Lightning-AI/lightning/issues/15428

File: src/lightning_app/utilities/load_app.py [hunk 2]
```diff
@@ -16,7 +16,31 @@
...
+    ] + listing
+    logger.error("".join(listing))
+    sys.exit(1)
+
+
+def load_app_from_file(filepath: str, raise_exception: bool = False) -> "LightningApp":
+    """Load a LightningApp from a file.
+
+    :param filepath:  The path to the file containing the LightningApp.
+    :param raise_exception: If True, raise an exception if the app cannot be loaded.
```
Comment (awaelchli) [inline src/lightning_app/utilities/load_app.py:41]: Let's follow the same doc style we already have

File: src/lightning_app/utilities/load_app.py [hunk 3]
```diff
@@ -16,7 +16,31 @@
 logger = Logger(__name__)
 
 
-def load_app_from_file(filepath: str) -> "LightningApp":
+def _prettifiy_exception(filepath: str):
+    """Pretty print the exception that occurred when loading the app.
+
+    :param filepath:
+    :return:
```
Comment (awaelchli) [inline src/lightning_app/utilities/load_app.py:23]:
```suggestion
```

File: src/lightning_app/runners/runtime.py [hunk 4]
```diff
@@ -151,3 +151,8 @@ def _add_stopped_status_to_work(self, work: "lightning_app.LightningWork") -> No
         latest_call_hash = work._calls[CacheCallsKeys.LATEST_CALL_HASH]
         if latest_call_hash in work._calls:
             work._calls[latest_call_hash]["statuses"].append(make_status(WorkStageStatus.STOPPED))
+
+    @classmethod
+    def load_app_from_file(cls, filepath: str) -> "LightningApp":
+
+        return load_app_from_file(filepath)
```
Comment (awaelchli) [inline src/lightning_app/runners/runtime.py:158]: why this indirection? can we update usages and call the function directly?

**Review by awaelchli — APPROVED** [review]

File: src/lightning_app/runners/cloud.py [hunk 5]
```diff
@@ -463,6 +465,30 @@ def _project_has_sufficient_credits(self, project: V1Membership, app: Optional[L
...
         return balance >= 1
 
+    @classmethod
+    def load_app_from_file(cls, filepath: str) -> "LightningApp":
+        """This is meant to use only locally for cloud runtime."""
+        try:
+            app = load_app_from_file(filepath, raise_exception=True)
+        except ModuleNotFoundError:
+            # this is very generic exception.
+            logger.info("Could not load the app() locally. Starting the app directly on the cloud.")
```
Comment (lantiga) [inline src/lightning_app/runners/cloud.py:475]:
```suggestion
            logger.info("Could not load the app locally. Starting the app directly on the cloud.")
```

**Review by hhsecond — APPROVED** [review]

**Pull request merged**